Upgrade to Bugsnag 4

Pezmc opened the pull request:
Upgrade ember-cli-bugsnag to use the latest 4.x version of the [bugsnag-js](https://github.com/bugsnag/bugsnag-js) client.

I followed their migration guide at https://github.com/bugsnag/bugsnag-js/blob/master/UPGRADING.md. There are two _important_ changes here, ember-cli-bugsnag now instantiates bugsnag, rather than configuring it, and we're using the new notify API to send along the metadata/user, rather than setting those before sending the error across.

As it's quite outdated, we also should upgrade the ember-source version and some of our other dependencies, but these can be done separately to keep the scope of the PR down.

Tests are green, but we should probably do some sanity QA!

Fixes #81, thanks to @alonski for laying some groundwork in https://github.com/binhums/ember-cli-bugsnag/pull/82 and https://github.com/binhums/ember-cli-bugsnag/pull/80!

Pezmc commented:
@binhums Do you have time to review this? I'd love to get it merged!

Alonski commented:
This looks really awesome @Pezmc 
I can test this on a staging server to see if this still works as it should and documents bugs

binhums commented:
Hi @Pezmc! Thank you so much for your efforts here. I regret that I'm not able to give this repo the attention it deserves. I've just finished a big project and so I should have a chance to look at this properly over the weekend.

If you're interested in helping maintain the project I'd be happy to discuss, could you email me using the email address on my profile?

binhums commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ember-cli-bugsnag",
-  "version": "1.3.0",
+  "version": "2.0.0",
```
Please leave out the version bump because we will do this as part of the deploy process. Agreed this should be a 'major' version change.

binhums commented on vendor/bugsnag/shim.js:
```diff
@@ -1,11 +1,11 @@
   if (typeof FastBoot === 'undefined') {
     return {
-      default: Bugsnag
+      default: bugsnag
```
could this be `window.bugsnag`

Pezmc replied:
Yes, yes it should

binhums commented:
At a first glance this looks good. I need to rationalise my feelings about using window.bugsnag instead of an imported library. Mostly I want to make sure that the library is loaded and configured at the right point to pick up errors when Ember is starting up.

@Alonski this will definitely need some manual QA - that would be lovely. I'd be happy to go through what we can do together if that helps.

Again - thank you both @Alonski and @Pezmc.

Pezmc commented:
Thanks for the review, I wasn't entirely happy about `window.bugsnag` being an instance either, but that's how they suggest you implement it.

We _could_ do some shuffling and export a configured instance of BugsnagClient so we could still import it? That feels nicer. Though that wouldn't guarantee it has loaded any more than in the current state.

We've been using this branch in production and tracked ~3.6k errors through it without issue, but I'm in favour of some extra manual QA!

I'll send you an email regarding donating some time here to keep the project maintained!

Pezmc commented:
I've dropped the version number bump (previously f7398ad) and updated `shim` to use `window.bugsnag` (exported as `global.bugsnag`) by [their client](https://github.com/bugsnag/bugsnag-js).

Pezmc commented:
What are your thoughts on the best way to pass around the instance of BugsnagClient avoiding the use of `window`? Chatting on the ember slack, someone suggested moving all the initialization into the instance-initializer.

jasonmevans commented:
Any chance this can be moved forward? 

I would be interested in helping maintain this project, as well. @binhums I didn't see your email address in your profile, so I wasn't able to contact you directly about it. Let me know!

pdud commented:
@jasonmevans thanks for working on this!

Unfortunately perhaps it's time to see if Adopted Ember Addons can provide some advice?

https://github.com/adopted-ember-addons?type=source

NPM does let you claim abandoned packages too if need be.

https://www.npmjs.com/policies/disputes

Alonski commented:
@binhums Would you be up for moving this repo and the npm package to Adopted Ember Addons?
If so you can transfer the repo and the package to me and I will get the rest sorted out

If someone wants to help out in maintaining ping me on Discord in the #adopted-ember-addons channel :)

cc @Pezmc @pdud @jasonmevans

binhums commented:
Hey friends, sorry about the radio silence - I appreciate I've been less than helpful. :pray:

I think transferring to adopted ember addons is an excellent idea. I would however like to manage the process myself.

If any of you would like to talk directly I would be happy to set up a call. You can contact me through the email address I have now added to my GitHub profile.

binhums commented:
Transfer complete.

Pezmc commented:
To move forward, I'd be on board with using `window.busnag` for now, so we at least have Bugsnag 4 support, then looking into shimming and reporting as a follow up!

Alonski commented:
@Pezmc Want to get this merged in? There are some conflicts :)

pdud commented:
Hey @Pezmc thank you so much for this PR!

Would you be interested in resolving the conflicts for us? :)

Pezmc commented:
Two years on, I don't work with Ember or BugSnag anymore, but I can likely easily resolve the yarn.lock conflict if that's what's holding this PR up!

Pezmc commented:
Rebased this branch onto master and resolved the conflicts in `package.json` and `yarn.lock`.

However, at this point, bugsnag-js is on version 7+, and the path from 4 to 7 contains many breaking changes. See: https://github.com/bugsnag/bugsnag-js/blob/master/UPGRADING.md#4x-to-5x

At this point, it may be worth someone restarting this PR from scratch with the latest build.

pdud commented:
Thanks @Pezmc!

I did notice it was on v7 too 😅 . I thought getting this in would set us on the right path for continuing that upgrade.

@jderr-mx looks like we can get this one in and work towards the latest bugsnag-js releases!

jderr-mx commented:
@pdud, @Alonski, @Pezmc I'll take a look. Looks good, I'll see about getting this released!

Alonski commented:
Great thanks @jderr-mx

jderr-mx commented:
Merging this in and releasing
Thanks @Pezmc, @pdud , and @Alonski !
